File: backend/src/main/java/com/sms/service/AdminService.java
package com.sms.service;

import com.sms.dto.*;
import com.sms.entity.Admin;
import com.sms.entity.Student;
import com.sms.repository.AdminRepository;
import com.sms.repository.StudentRepository;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class AdminService {

    private final AdminRepository adminRepository;
    private final StudentRepository studentRepository;
    private final PasswordEncoder passwordEncoder;

    public AdminService(AdminRepository adminRepository,
                        StudentRepository studentRepository,
                        PasswordEncoder passwordEncoder) {
        this.adminRepository = adminRepository;
        this.studentRepository = studentRepository;
        this.passwordEncoder = passwordEncoder;
    }

    // --- Admin profile ---

    public ProfileResponse getProfile(String email) {
        Admin admin = adminRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Admin not found"));
        return toProfileResponse(admin);
    }

    public ProfileResponse updateProfile(String email, ProfileUpdateRequest request) {
        Admin admin = adminRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Admin not found"));
        if (request.getName() != null) admin.setName(request.getName());
        if (request.getDescription() != null) admin.setDescription(request.getDescription());
        admin = adminRepository.save(admin);
        return toProfileResponse(admin);
    }

    // --- Student management ---

    public List<ProfileResponse> getAllStudents() {
        return studentRepository.findAll().stream()
                .map(this::toProfileResponse)
                .collect(Collectors.toList());
    }

    public ProfileResponse addStudent(StudentCreateRequest request) {
        if (studentRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Student with this email already exists");
        }
        if (adminRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("This email is already used by an admin");
        }

        Student student = new Student(
                request.getEmail(),
                passwordEncoder.encode(request.getPassword()),
                request.getName()
        );
        student = studentRepository.save(student);
        return toProfileResponse(student);
    }

    public ProfileResponse updateStudent(Long studentId, StudentUpdateRequest request) {
        Student student = studentRepository.findById(studentId)
                .orElseThrow(() -> new RuntimeException("Student not found"));

        if (request.getEmail() != null) {
            if (!request.getEmail().equals(student.getEmail()) &&
                    (studentRepository.existsByEmail(request.getEmail()) ||
                     adminRepository.existsByEmail(request.getEmail()))) {
                throw new RuntimeException("Email is already in use");
            }
            student.setEmail(request.getEmail());
        }
        if (request.getPassword() != null) {
            student.setPassword(passwordEncoder.encode(request.getPassword()));
        }
        student = studentRepository.save(student);
        return toProfileResponse(student);
    }

    public void removeStudent(Long studentId) {
        if (!studentRepository.existsById(studentId)) {
            throw new RuntimeException("Student not found");
        }
        studentRepository.deleteById(studentId);
    }

    // --- Admin management ---

    public List<ProfileResponse> getAllAdmins() {
        return adminRepository.findAll().stream()
                .map(this::toProfileResponse)
                .collect(Collectors.toList());
    }

    public ProfileResponse addAdmin(AdminCreateRequest request) {
        if (adminRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("Admin with this email already exists");
        }
        if (studentRepository.existsByEmail(request.getEmail())) {
            throw new RuntimeException("This email is already used by a student");
        }

        Admin admin = new Admin(
                request.getEmail(),
                passwordEncoder.encode(request.getPassword()),
                request.getName()
        );
        admin = adminRepository.save(admin);
        return toProfileResponse(admin);
    }

    public void removeAdmin(Long adminId, String currentAdminEmail) {
        Admin target = adminRepository.findById(adminId)
                .orElseThrow(() -> new RuntimeException("Admin not found"));
        if (target.getEmail().equals(currentAdminEmail)) {
            throw new RuntimeException("Cannot remove yourself");
        }
        adminRepository.deleteById(adminId);
    }

    // --- Helpers ---

    private ProfileResponse toProfileResponse(Admin admin) {
        return new ProfileResponse(admin.getId(), admin.getEmail(), admin.getName(),
                admin.getDescription(), admin.getCreatedAt(), admin.getUpdatedAt());
    }

    private ProfileResponse toProfileResponse(Student student) {
        return new ProfileResponse(student.getId(), student.getEmail(), student.getName(),
                student.getDescription(), student.getCreatedAt(), student.getUpdatedAt());
    }
}

File: backend/app/services/admin_service.py
from app import db
from app.models import Admin, Student
from app.services.auth_service import hash_password


def get_profile(email):
    admin = Admin.query.filter_by(email=email).first()
    if not admin:
        raise ValueError("Admin not found")
    return admin.to_profile()


def update_profile(email, data):
    admin = Admin.query.filter_by(email=email).first()
    if not admin:
        raise ValueError("Admin not found")

    if "name" in data and data["name"] is not None:
        admin.name = data["name"]
    if "description" in data and data["description"] is not None:
        admin.description = data["description"]

    db.session.commit()
    return admin.to_profile()


def get_all_students():
    students = Student.query.all()
    return [s.to_profile() for s in students]


def add_student(data):
    email = data.get("email")
    password = data.get("password")
    name = data.get("name")

    if not email or not password:
        raise ValueError("Email and password are required")

    if Student.query.filter_by(email=email).first():
        raise ValueError("Email already exists")
    if Admin.query.filter_by(email=email).first():
        raise ValueError("Email already exists")

    student = Student(
        email=email,
        password=hash_password(password),
        name=name,
    )
    db.session.add(student)
    db.session.commit()
    return student.to_profile()


def update_student(student_id, data):
    student = db.session.get(Student, student_id)
    if not student:
        raise ValueError("Student not found")

    if "email" in data and data["email"] is not None:
        new_email = data["email"]
        if new_email != student.email:
            if Student.query.filter_by(email=new_email).first():
                raise ValueError("Email already exists")
            if Admin.query.filter_by(email=new_email).first():
                raise ValueError("Email already exists")
            student.email = new_email

    if "password" in data and data["password"] is not None:
        student.password = hash_password(data["password"])

    db.session.commit()
    return student.to_profile()


def remove_student(student_id):
    student = db.session.get(Student, student_id)
    if not student:
        raise ValueError("Student not found")
    db.session.delete(student)
    db.session.commit()


def get_all_admins():
    admins = Admin.query.all()
    return [a.to_profile() for a in admins]


def add_admin(data):
    email = data.get("email")
    password = data.get("password")
    name = data.get("name")

    if not email or not password:
        raise ValueError("Email and password are required")

    if Admin.query.filter_by(email=email).first():
        raise ValueError("Email already exists")
    if Student.query.filter_by(email=email).first():
        raise ValueError("Email already exists")

    admin = Admin(
        email=email,
        password=hash_password(password),
        name=name,
    )
    db.session.add(admin)
    db.session.commit()
    return admin.to_profile()


def remove_admin(admin_id, current_admin_email):
    admin = db.session.get(Admin, admin_id)
    if not admin:
        raise ValueError("Admin not found")
    if admin.email == current_admin_email:
        raise ValueError("Cannot delete yourself")
    db.session.delete(admin)
    db.session.commit()
